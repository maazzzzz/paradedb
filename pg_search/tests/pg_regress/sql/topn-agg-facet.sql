-- Test TopN + Aggregates + Faceting
-- Phase 1: Basic TopN tests with window aggregate detection

CREATE EXTENSION IF NOT EXISTS pg_search;

DROP TABLE IF EXISTS products CASCADE;

-- Setup test data
CREATE TABLE products (
    id SERIAL PRIMARY KEY,
    name TEXT,
    description TEXT,
    category TEXT,
    brand TEXT,
    price NUMERIC,
    rating NUMERIC,
    in_stock BOOLEAN,
    sales INTEGER
);

-- Insert test data
INSERT INTO products (name, description, category, brand, price, rating, in_stock, sales) VALUES
    ('MacBook Pro', 'High-performance laptop for professionals', 'Laptops', 'Apple', 2499, 4.8, true, 150),
    ('Dell XPS 13', 'Compact and powerful ultrabook', 'Laptops', 'Dell', 1299, 4.6, true, 200),
    ('ThinkPad X1', 'Business laptop with great keyboard', 'Laptops', 'Lenovo', 1599, 4.5, true, 180),
    ('HP Spectre', 'Stylish convertible laptop', 'Laptops', 'HP', 1399, 4.4, true, 120),
    ('ASUS ROG', 'Gaming laptop with RTX graphics', 'Laptops', 'ASUS', 1899, 4.7, true, 90);

-- Create BM25 index
CREATE INDEX products_idx ON products
USING bm25(id, name, description, category, brand, price, rating, in_stock, sales)
WITH (
    key_field='id',
    text_fields='{
        "name": {},
        "description": {},
        "brand": {"fast": true}
    }',
    numeric_fields='{
        "price": {"fast": true},
        "rating": {"fast": true},
        "sales": {"fast": true}
    }',
    boolean_fields='{
        "in_stock": {"fast": true}
    }'
);

-- Test 1: Basic TopN without window aggregates
\echo 'Test 1: Basic TopN query'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    category,
    rating
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

SELECT 
    id,
    name,
    category,
    rating
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 2: TopN with COUNT(*) OVER () - Basic window aggregate
\echo 'Test 2: TopN with COUNT(*) OVER () (basic window aggregate)'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    COUNT(*) OVER () as total_count
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

SELECT 
    id,
    name,
    rating,
    COUNT(*) OVER () as total_count
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 3: Multiple window aggregates (COUNT, SUM, AVG)
\echo 'Test 3: Multiple window aggregates in one query'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    price,
    COUNT(*) OVER () as total_count,
    SUM(price) OVER () as total_price,
    AVG(rating) OVER () as avg_rating
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

SELECT 
    id,
    name,
    rating,
    price,
    COUNT(*) OVER () as total_count,
    SUM(price) OVER () as total_price,
    AVG(rating) OVER () as avg_rating
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 4: Window aggregate with PARTITION BY (not supported yet, but should extract)
\echo 'Test 4: COUNT with PARTITION BY category'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    category,
    rating,
    COUNT(*) OVER (PARTITION BY category) as category_count
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 5: Window aggregate with ORDER BY in OVER clause
\echo 'Test 5: SUM with ORDER BY in OVER clause (running total)'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    price,
    SUM(price) OVER (ORDER BY rating DESC) as running_total
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 6: Window aggregate with ROWS frame
\echo 'Test 6: AVG with ROWS frame'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    AVG(rating) OVER (ORDER BY rating DESC ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING) as moving_avg
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 7: MIN and MAX window aggregates
\echo 'Test 7: MIN and MAX aggregates'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    price,
    MIN(price) OVER () as min_price,
    MAX(price) OVER () as max_price
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

SELECT 
    id,
    name,
    rating,
    price,
    MIN(price) OVER () as min_price,
    MAX(price) OVER () as max_price
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 8: Window aggregate with FILTER clause
\echo 'Test 8: COUNT with FILTER clause'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    in_stock,
    COUNT(*) FILTER (WHERE in_stock = true) OVER () as in_stock_count
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 9: COUNT with specific column (not *)
\echo 'Test 9: COUNT(column) instead of COUNT(*)'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    COUNT(brand) OVER () as brand_count
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 10: Complex PARTITION BY and ORDER BY combination
\echo 'Test 10: Complex window with PARTITION BY and ORDER BY'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    category,
    brand,
    rating,
    COUNT(*) OVER (PARTITION BY category ORDER BY rating DESC) as category_rank_count
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 11: Window aggregate without ORDER BY on base query
\echo 'Test 11: Window aggregate without ORDER BY or LIMIT'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    COUNT(*) OVER () as total_count
FROM products
WHERE description @@@ 'laptop';

SELECT 
    id,
    name,
    rating,
    COUNT(*) OVER () as total_count
FROM products
WHERE description @@@ 'laptop';

-- Test 12: Window aggregate with RANGE frame
\echo 'Test 12: SUM with RANGE frame'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    price,
    SUM(price) OVER (ORDER BY rating RANGE BETWEEN 0.5 PRECEDING AND 0.5 FOLLOWING) as range_sum
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 13: Multiple different PARTITION BY clauses
\echo 'Test 13: Multiple window functions with different partitions'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    category,
    brand,
    rating,
    COUNT(*) OVER (PARTITION BY category) as by_category,
    COUNT(*) OVER (PARTITION BY brand) as by_brand,
    COUNT(*) OVER () as total
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 14: Window aggregate with GROUPS frame (PG17+)
\echo 'Test 14: COUNT with GROUPS frame'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    COUNT(*) OVER (ORDER BY rating GROUPS BETWEEN 1 PRECEDING AND CURRENT ROW) as group_count
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;

-- Test 15: TopN with no @@@ operator (should not trigger window function handling)
\echo 'Test 15: Window aggregate without @@@ operator (should use standard WindowAgg)'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    COUNT(*) OVER () as total_count
FROM products
WHERE rating > 4.5
ORDER BY rating DESC
LIMIT 3;

SELECT 
    id,
    name,
    rating,
    COUNT(*) OVER () as total_count
FROM products
WHERE rating > 4.5
ORDER BY rating DESC
LIMIT 3;

-- Test 16: Window aggregate with multiple base table columns
\echo 'Test 16: Window aggregate with all result columns'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    category,
    brand,
    price,
    rating,
    sales,
    COUNT(*) OVER () as total_count,
    SUM(sales) OVER () as total_sales
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 2;

SELECT 
    id,
    name,
    category,
    brand,
    price,
    rating,
    sales,
    COUNT(*) OVER () as total_count,
    SUM(sales) OVER () as total_sales
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 2;

-- Test 17: Window aggregate in a subquery (TopN in outer query)
\echo 'Test 17: Subquery with window aggregate, outer LIMIT'
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT * FROM (
    SELECT 
        id,
        name,
        rating,
        price,
        COUNT(*) OVER () as total_count
    FROM products
    WHERE description @@@ 'laptop'
    ORDER BY rating DESC
    LIMIT 5
) sub
ORDER BY price DESC
LIMIT 2;

SELECT * FROM (
    SELECT 
        id,
        name,
        rating,
        price,
        COUNT(*) OVER () as total_count
    FROM products
    WHERE description @@@ 'laptop'
    ORDER BY rating DESC
    LIMIT 5
) sub
ORDER BY price DESC
LIMIT 2;

-- Cleanup
DROP TABLE products CASCADE;