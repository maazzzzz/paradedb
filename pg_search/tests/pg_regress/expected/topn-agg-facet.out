-- Test TopN + Aggregates + Faceting
-- Phase 1: Basic TopN tests with window aggregate detection
CREATE EXTENSION IF NOT EXISTS pg_search;
DROP TABLE IF EXISTS products CASCADE;
-- Setup test data
CREATE TABLE products (
    id SERIAL PRIMARY KEY,
    name TEXT,
    description TEXT,
    category TEXT,
    brand TEXT,
    price NUMERIC,
    rating NUMERIC,
    in_stock BOOLEAN,
    sales INTEGER
);
-- Insert test data
INSERT INTO products (name, description, category, brand, price, rating, in_stock, sales) VALUES
    ('MacBook Pro', 'High-performance laptop for professionals', 'Laptops', 'Apple', 2499, 4.8, true, 150),
    ('Dell XPS 13', 'Compact and powerful ultrabook', 'Laptops', 'Dell', 1299, 4.6, true, 200),
    ('ThinkPad X1', 'Business laptop with great keyboard', 'Laptops', 'Lenovo', 1599, 4.5, true, 180),
    ('HP Spectre', 'Stylish convertible laptop', 'Laptops', 'HP', 1399, 4.4, true, 120),
    ('ASUS ROG', 'Gaming laptop with RTX graphics', 'Laptops', 'ASUS', 1899, 4.7, true, 90);
-- Create BM25 index
CREATE INDEX products_idx ON products
USING bm25(id, name, description, category, brand, price, rating, in_stock, sales)
WITH (
    key_field='id',
    text_fields='{
        "name": {},
        "description": {},
        "brand": {"fast": true}
    }',
    numeric_fields='{
        "price": {"fast": true},
        "rating": {"fast": true},
        "sales": {"fast": true}
    }',
    boolean_fields='{
        "in_stock": {"fast": true}
    }'
);
-- Test 1: Basic TopN without window aggregates
\echo 'Test 1: Basic TopN query'
Test 1: Basic TopN query
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    category,
    rating
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;
WARNING:  query_has_search_operator: Looking for opno=937092 or 937085
WARNING:  query_has_search_operator: jointree is not null
WARNING:  query_has_search_operator: quals is not null, checking...
WARNING:  expr_contains_operator: Checking node type=T_OpExpr
WARNING:  expr_contains_operator: Found OpExpr with opno=937085
WARNING:  expr_contains_operator: MATCH! Found target operator
WARNING:  query_has_search_operator: Found @@@ in WHERE clause
WARNING:  planner_hook: SELECT query - has_window_funcs=false, has_search_op=true
WARNING:  plan_custom_path: Window aggregates NOT in storage
WARNING:  plan_custom_path: Failed to retrieve window aggregates
                                                                         QUERY PLAN                                                                          
-------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   Output: id, name, category, rating
   ->  Custom Scan (ParadeDB Scan) on public.products
         Output: id, name, category, rating
         Table: products
         Index: products_idx
         Exec Method: TopNScanExecState
         Scores: false
            TopN Order By: rating desc
            TopN Limit: 3
         Tantivy Query: {"with_index":{"query":{"parse_with_field":{"field":"description","query_string":"laptop","lenient":null,"conjunction_mode":null}}}}
(11 rows)

SELECT 
    id,
    name,
    category,
    rating
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;
WARNING:  query_has_search_operator: Looking for opno=937092 or 937085
WARNING:  query_has_search_operator: jointree is not null
WARNING:  query_has_search_operator: quals is not null, checking...
WARNING:  expr_contains_operator: Checking node type=T_OpExpr
WARNING:  expr_contains_operator: Found OpExpr with opno=937085
WARNING:  expr_contains_operator: MATCH! Found target operator
WARNING:  query_has_search_operator: Found @@@ in WHERE clause
WARNING:  planner_hook: SELECT query - has_window_funcs=false, has_search_op=true
WARNING:  plan_custom_path: Window aggregates NOT in storage
WARNING:  plan_custom_path: Failed to retrieve window aggregates
WARNING:  inject_score_and_snippet_placeholders: need_scores=false, need_snippets=false, has_window_aggs=false
 id |    name     | category | rating 
----+-------------+----------+--------
  1 | MacBook Pro | Laptops  |    4.8
  5 | ASUS ROG    | Laptops  |    4.7
  3 | ThinkPad X1 | Laptops  |    4.5
(3 rows)

-- Test 2: TopN with COUNT(*) OVER () - should trigger window aggregate detection
\echo 'Test 2: TopN with COUNT(*) OVER () (window aggregate detection)'
Test 2: TopN with COUNT(*) OVER () (window aggregate detection)
EXPLAIN (COSTS OFF, VERBOSE, TIMING OFF)
SELECT 
    id,
    name,
    rating,
    COUNT(*) OVER () as total_count
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;
WARNING:  query_has_search_operator: Looking for opno=937092 or 937085
WARNING:  query_has_search_operator: jointree is not null
WARNING:  query_has_search_operator: quals is not null, checking...
WARNING:  expr_contains_operator: Checking node type=T_OpExpr
WARNING:  expr_contains_operator: Found OpExpr with opno=937085
WARNING:  expr_contains_operator: MATCH! Found target operator
WARNING:  query_has_search_operator: Found @@@ in WHERE clause
WARNING:  planner_hook: SELECT query - has_window_funcs=true, has_search_op=true
WARNING:  planner_hook: Found window functions in search query, extracting and storing
WARNING:  Extracted window spec: partition_by=[], order_by=None, frame_clause=None
WARNING:  ═══════════════════════════════════════════════════════════
WARNING:  EXTRACTED WINDOW FUNCTION #3
WARNING:  ═══════════════════════════════════════════════════════════
WARNING:  Aggregate Type: COUNT(*)
WARNING:  PARTITION BY: (none - aggregate over entire result set)
WARNING:  ORDER BY: (none)
WARNING:  FRAME: (default)
WARNING:  Result Type OID: 20
WARNING:  Target Entry Index: 3
WARNING:  ✅ CAN EXECUTE: Yes - will compute at execution time
WARNING:  ═══════════════════════════════════════════════════════════
WARNING:  planner_hook: Extracted 1 window aggregates
WARNING:  planner_hook: Stored window aggregates in global storage
WARNING:  plan_custom_path: Window aggregates ARE in storage, retrieving...
WARNING:  Retrieved window aggregates from global storage
WARNING:  plan_custom_path: Retrieved 1 window aggregates from global storage
WARNING:  Initializing scan state with 1 window aggregates
                                                                               QUERY PLAN                                                                                
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit
   Output: id, name, rating, (count(*) OVER (?))
   ->  Sort
         Output: id, name, rating, (count(*) OVER (?))
         Sort Key: products.rating DESC
         ->  WindowAgg
               Output: id, name, rating, count(*) OVER (?)
               ->  Custom Scan (ParadeDB Scan) on public.products
                     Output: id, name, rating
                     Table: products
                     Index: products_idx
                     Exec Method: NormalScanExecState
                     Scores: false
                     Tantivy Query: {"with_index":{"query":{"parse_with_field":{"field":"description","query_string":"laptop","lenient":null,"conjunction_mode":null}}}}
(14 rows)

SELECT 
    id,
    name,
    rating,
    COUNT(*) OVER () as total_count
FROM products
WHERE description @@@ 'laptop'
ORDER BY rating DESC
LIMIT 3;
WARNING:  query_has_search_operator: Looking for opno=937092 or 937085
WARNING:  query_has_search_operator: jointree is not null
WARNING:  query_has_search_operator: quals is not null, checking...
WARNING:  expr_contains_operator: Checking node type=T_OpExpr
WARNING:  expr_contains_operator: Found OpExpr with opno=937085
WARNING:  expr_contains_operator: MATCH! Found target operator
WARNING:  query_has_search_operator: Found @@@ in WHERE clause
WARNING:  planner_hook: SELECT query - has_window_funcs=true, has_search_op=true
WARNING:  planner_hook: Found window functions in search query, extracting and storing
WARNING:  Extracted window spec: partition_by=[], order_by=None, frame_clause=None
WARNING:  ═══════════════════════════════════════════════════════════
WARNING:  EXTRACTED WINDOW FUNCTION #3
WARNING:  ═══════════════════════════════════════════════════════════
WARNING:  Aggregate Type: COUNT(*)
WARNING:  PARTITION BY: (none - aggregate over entire result set)
WARNING:  ORDER BY: (none)
WARNING:  FRAME: (default)
WARNING:  Result Type OID: 20
WARNING:  Target Entry Index: 3
WARNING:  ✅ CAN EXECUTE: Yes - will compute at execution time
WARNING:  ═══════════════════════════════════════════════════════════
WARNING:  planner_hook: Extracted 1 window aggregates
WARNING:  planner_hook: Stored window aggregates in global storage
WARNING:  plan_custom_path: Window aggregates ARE in storage, retrieving...
WARNING:  Retrieved window aggregates from global storage
WARNING:  plan_custom_path: Retrieved 1 window aggregates from global storage
WARNING:  Initializing scan state with 1 window aggregates
WARNING:  inject_score_and_snippet_placeholders: need_scores=false, need_snippets=false, has_window_aggs=true
 id |    name     | rating | total_count 
----+-------------+--------+-------------
  1 | MacBook Pro |    4.8 |           4
  5 | ASUS ROG    |    4.7 |           4
  3 | ThinkPad X1 |    4.5 |           4
(3 rows)

-- Cleanup
DROP TABLE products CASCADE;
